module/zfs/qat_compress.c

```diff
@@ -0,0 +1,528 @@
+				 *  size here.
+				 */
+				status = PHYS_CONTIG_ALLOC(
+				    &buffer_array[inst][buff_num]->pBuffers->
+				    pData, 2 * QAT_MAX_BUF_SIZE);
+				buffer_array[inst][buff_num]->numBuffers = 1;
+				buffer_array[inst][buff_num]->pBuffers->
+				    dataLenInBytes = 2 * QAT_MAX_BUF_SIZE;
+			}
+		}
+
+		status = cpaDcStartInstance(dc_inst_handles[inst],
+		    num_inter_buff_lists, buffer_array[inst]);
+		if (status != CPA_STATUS_SUCCESS)
+			goto fail;
+
+		sd.compLevel = CPA_DC_L1;
+		sd.compType = CPA_DC_DEFLATE;
+		sd.huffType = CPA_DC_HT_FULL_DYNAMIC;
+		sd.sessDirection = CPA_DC_DIR_COMBINED;
+		sd.sessState = CPA_DC_STATELESS;
+		sd.deflateWindowSize = 7;
+		sd.checksum = CPA_DC_ADLER32;
+		status = cpaDcGetSessionSize(dc_inst_handles[inst],
+		    &sd, &sess_size, &ctx_size);
+		if (status != CPA_STATUS_SUCCESS)
+			goto fail;
+
+		PHYS_CONTIG_ALLOC(&session_handles[inst], sess_size);
+		if (session_handles[inst] == NULL)
+			goto fail;
+
+		status = cpaDcInitSession(dc_inst_handles[inst],
+		    session_handles[inst],
+		    &sd, NULL, qat_dc_callback);
+		if (status != CPA_STATUS_SUCCESS)
+			goto fail;
+	}
+	qat_ksp = kstat_create("zfs", 0, "qat", "misc",
+	    KSTAT_TYPE_NAMED, sizeof (qat_stats) / sizeof (kstat_named_t),
+	    KSTAT_FLAG_VIRTUAL);
+
+	if (qat_ksp != NULL) {
+		qat_ksp->ks_data = &qat_stats;
+		kstat_install(qat_ksp);
+	}
+
+	qat_init_done = 1;
+	return (0);
+fail:
+	qat_clean();
+	return (-1);
+}
+
+void
+qat_fini(void)
+{
+	qat_clean();
+
+	if (qat_ksp != NULL) {
+		kstat_delete(qat_ksp);
+		qat_ksp = NULL;
+	}
+}
+
+int
+use_qat(size_t s_len)
+{
+	if (zfs_qat_disable == B_TRUE ||
+	    qat_init_done == 0 ||
+	    s_len < QAT_MIN_BUF_SIZE ||
+	    s_len > QAT_MAX_BUF_SIZE) {
+		return (0);
+	}
+	return (1);
+}
+
+int
+qat_compress(int dir, char *src, int src_len,
+    char *dst, int dst_len, size_t *c_len)
+{
+	CpaInstanceHandle dc_inst_handle;
+	CpaDcSessionHandle session_handle;
+	CpaBufferList *buf_list_src = NULL;
+	CpaBufferList *buf_list_dst = NULL;
+	CpaFlatBuffer *flat_buf_src = NULL;
+	CpaFlatBuffer *flat_buf_dst = NULL;
+	Cpa8U *buffer_meta_src = NULL;
+	Cpa8U *buffer_meta_dst = NULL;
+	Cpa32U buffer_meta_size = 0;
+	CpaDcRqResults dc_results;
+	CpaStatus status = CPA_STATUS_SUCCESS;
+	Cpa32U hdr_sz = 0;
+	Cpa32U compressed_sz;
+	Cpa32U num_src_buf = (src_len >> PAGE_SHIFT) + 1;
+	Cpa32U num_dst_buf = (dst_len >> PAGE_SHIFT) + 1;
+	Cpa32U bytes_left;
+	char *data;
+	struct page *in_page, *out_page;
+	struct page **in_pages = NULL;
+	struct page **out_pages = NULL;
+	struct completion complete;
+	size_t ret = -1;
+	int page_num = 0;
+	int i;
+
+	Cpa32U src_buffer_list_mem_size = sizeof (CpaBufferList) +
+	    (num_src_buf * sizeof (CpaFlatBuffer));
+	Cpa32U dst_buffer_list_mem_size = sizeof (CpaBufferList) +
+	    (num_dst_buf * sizeof (CpaFlatBuffer));
+
+	if (!is_vmalloc_addr(src) || !is_vmalloc_addr(src + src_len - 1) ||
+	    !is_vmalloc_addr(dst) || !is_vmalloc_addr(dst + dst_len - 1))
```

Comment:
We got problem when using kmap to map the page to physical address if it is not vmalloced, so we only accept the vmalloc memory so far. As Brian motioned, once the ABD changes can provide a vector of pages to compressor, we can adjust the code here then.